fix(issue #2205): FK is already defined as the defined field

TcMits opened the pull request:
Hello @a8m

Fixed #2205

a8m reviewed (commented):
Hey @TcMits, and thanks for the contribution. 

I want to review this but there are too many unrelated changes in this PR. Please, clean them and tag me when it's ready. Thanks 🙏

TcMits commented:
Hey @a8m , i removed unrelated changes. Sorry i forgot turning off auto formatter. 😢

TcMits commented:
Hey @a8m, i removed unrelated changes. it's ready to review. Thanks 😄

a8m commented on entc/gen/type.go:
```diff
@@ -708,16 +708,33 @@ func (t *Type) setupFKs() error {
-		if key, _ := e.StorageKey(); key != nil && len(key.Columns) == 1 && key.Columns[0] == refid.StorageKey() {
-			fk.Field = refid
-			fk.UserDefined = true
+		// Special case for checking if the FK is already defined as the defined field (see issue 1288, 2205).
```
Please, add a test case to help understand what this change does. Thanks

TcMits replied:
Hi @a8m, i added a test case for this. Hope it helps 😄

TcMits replied:
Hey @a8m, i added a test case for this. Hope it helps 😄

a8m replied:
Thanks, but integration tests are also necessary (see `entc/integration`).

TcMits replied:
added @a8m. 😄

a8m commented on entc/integration/edgefield/ent/schema/user.go:
```diff
@@ -36,6 +36,7 @@ func (User) Fields() []ent.Field {
 func (User) Edges() []ent.Edge {
 	return []ent.Edge{
 		edge.To("pets", Pet.Type),
+		edge.To("previous_pets", Pet.Type).StorageKey(edge.Column("previous_owner_id")),
```
Why is StorageKey defined here? It should be defined on the schema that holds the foreign key using `Field` instead. Ent should throw an error in this case. 

Do you have any other case that reproduces it?

a8m commented:
Hey! Please, see my comment on [#2205](https://github.com/ent/ent/issues/2205#issuecomment-1295976877) and let me know if there's anything else that this PR solves or if you think this behavior is necessary (for detecting the reference in the "other" side).

TcMits commented:
Hey @a8m, 

You're right, throwing error in this case ( #2205 ) is the fastest way to solve problem.  But following the [docs](https://entgo.io/docs/schema-edges#storagekey):
> By default, Ent configures edge storage-keys by the edge-owner (the schema that holds the edge.To), and not the by back-reference (edge.From). This is because back-references are optional and can be removed.

So i think in most cases, we should define `storage-keys` in `edge-owner`. Then we can remove back-reference without any problems ( affect relations ). Sorry if i missed somethings. 😄

I tried default behavior of Ent when i removed back-references ( in edgefield integration test ), output struct of `pet.go` file:
```go
type Pet struct {
        ...
	PreviousOwnerID int `json:"previous_owner_id,omitempty"`
        ...
	previous_owner_id *int // didn't expect this
}
```
I added a [commit](https://github.com/ent/ent/pull/3032/commits/68035817585965610f0f8bd10e4119aebf436b2a) to change default behavior of Ent. Please tell me if i'm wrong.

Thanks.

a8m commented:
Understand your point @TcMits, and it's appreciated. But if the edge field is defined in the referenced schema, I expect the users to use it / declare it there. I find it cleaner, more explicit, and works well with other integrations, such as EntGQL.

a8m commented:
Again, your contribution to Ent is highly appreciated. I see it a bit differently when it comes to the schema definition.

TcMits commented:
> Again, your contribution to Ent is highly appreciated. I see it a bit differently when it comes to the schema definition.

Thanks. @a8m 

> Understand your point @TcMits, and it's appreciated. But if the edge field is defined in the referenced schema, I expect the users to use it / declare it there. I find it cleaner, more explicit, and works well with other integrations, such as EntGQL.

Okay, i will close this PR.

a8m commented:
❤️
